packages/aws-cdk-lib/core/lib/stack-synthesizers/asset-manifest-builder.ts

```diff
@@ -174,10 +175,24 @@ export class AssetManifestBuilder {
   }
 
   private manifestEnvName(stack: Stack): string {
-    return [
-      resolvedOr(stack.account, 'current_account'),
-      resolvedOr(stack.region, 'current_region'),
-    ].join('-');
+    const account = resolvedOr(stack.account, 'current_account');
+    const region = resolvedOr(stack.region, 'current_region');
+
+    // Create a hash of all destination-relevant properties to ensure uniqueness
+    // This ensures assets with same content but different destinations are published separately
+    const destinationProps = {
+      account,
+      region,
+      stackName: stack.stackName,
```

Comment:
Mixing in the `stackName` will make it so that assets that are duplicated between stacks *must* be uploaded twice. That is a stronger guarantee than we need. Try the actual destination's properties instead.